Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp

```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
-    parameters.topContentInset = 0;
-    m_process->send(Messages::WebProcess::CreateWebPage(m_pageID, parameters), 0);
-    m_process->addVisitedLinkStoreUser(page.visitedLinkStore(), page.identifier());
-
-    drawingArea->attachToProvisionalFrameProcess(m_process);
+    if (newProcess) {
+        auto parameters = page.creationParameters(m_process, *drawingArea);
+        parameters.subframeProcessFrameTreeInitializationParameters = { {
+            frame.frameID(),
+            *page.frameTreeCreationParameters(),
+            m_layerHostingContextIdentifier
+        } };
+        parameters.isProcessSwap = true; // FIXME: This should be a parameter to creationParameters rather than doctoring up the parameters afterwards.
+        parameters.topContentInset = 0;
+        m_process->send(Messages::WebProcess::CreateWebPage(m_pageID, parameters), 0);
+        m_process->addVisitedLinkStoreUser(page.visitedLinkStore(), page.identifier());
+        drawingArea->attachToProvisionalFrameProcess(m_process);
+    }
 
     LoadParameters loadParameters;
     loadParameters.request = request;
     loadParameters.shouldTreatAsContinuingLoad = WebCore::ShouldTreatAsContinuingLoad::YesAfterNavigationPolicyDecision;
     loadParameters.frameIdentifier = frame.frameID();
+    if (frame.parentFrame())
+        loadParameters.parentFrameIdentifier = frame.parentFrame()->frameID();
+    loadParameters.layerHostingContextIdentifier = m_layerHostingContextIdentifier;
     // FIXME: Add more parameters as appropriate.
 
     // FIXME: This gives too much cookie access. This should be removed after putting the entire frame tree in all web processes.
     auto giveAllCookieAccess = LoadedWebArchive::Yes;
-    page.websiteDataStore().networkProcess().sendWithAsyncReply(Messages::NetworkProcess::AddAllowedFirstPartyForCookies(m_process->coreProcessIdentifier(), WebCore::RegistrableDomain(request.url()), giveAllCookieAccess), [process = m_process, loadParameters = WTFMove(loadParameters), pageID = m_pageID] () mutable {
-        // FIXME: Do we need a LoadRequestWaitingForProcessLaunch version?
-        process->send(Messages::WebPage::LoadRequest(loadParameters), pageID);
-    });
+    if (newProcess) {
+        page.websiteDataStore().networkProcess().sendWithAsyncReply(Messages::NetworkProcess::AddAllowedFirstPartyForCookies(m_process->coreProcessIdentifier(), WebCore::RegistrableDomain(request.url()), giveAllCookieAccess), [process = m_process, loadParameters = WTFMove(loadParameters), pageID = m_pageID] () mutable {
+            // FIXME: Do we need a LoadRequestWaitingForProcessLaunch version?
+            process->send(Messages::WebPage::LoadRequest(loadParameters), pageID);
+        });
+    } else {
+        m_process->send(Messages::WebPage::LoadRequestByCreatingNewLocalFrameOrConvertingRemoteFrame(loadParameters), m_pageID);
```

Comment:
This needs to not be in the constructor.

Reply:
based on offline convo, we can clean this up later if we think of a different design.